### PR TITLE
Fix #239 - interpolated action "update" don't get params "upsert" and…

### LINKED ISSUE #239: doc_as_upsert problem with dynamic 'action'
I think doc_as_upsert does not work properly in 1.0.7 (logstash 1.5.4, elastic 1.4.2 or 1.7.1)

I'm testing with this logstash.conf

```
input {
    stdin {
        add_field => {
            "type" => "type_1"
            "[@metadata][_id]" => "id_1"
            "[@metadata][action]" => "update"
        }
    }
}

filter {
}

output {
    elasticsearch {
        action => "%{[@metadata][action]}"
        document_id => "%{[@metadata][_id]}"
        doc_as_upsert => "true"
        host => "localhost"
        protocol => "http"
        index => "test_index"
    }
}
```

elastic is running, logstash is running by `logstash -f logstash.conf`

Let's remove the index first

```
curl -XDELETE http://localhost:9200/test_index
```

Then pass any text to logstash console window:

```
Logstash startup completed
Hello
failed action with response of 404, dropping action: ["update", {:_id=>"id_1", :_index=>"test_index", :_type=>"type_1", :_routing=>nil}, #<LogStash::Event:0x3a0d6f83 @metadata_accessors=#<LogStash::Util::Accessors:0x66f2a7ef @store={"_id"=>"id_1", "action"=>"update", "retry_count"=>0}, @lut={"[_id]"=>[{"_id"=>"id_1", "action"=>"update", "retry_count"=>0}, "_id"], "[action]"=>[{"_id"=>"id_1", "action"=>"update", "retry_count"=>0}, "action"]}>, @cancelled=false, @data={"message"=>"Hello", "@version"=>"1", "@timestamp"=>"2015-08-31T12:06:08.273Z", "type"=>"type_1", "host"=>"minix.local"}, @metadata={"_id"=>"id_1", "action"=>"update", "retry_count"=>0}, @accessors=#<LogStash::Util::Accessors:0x6718b5a2 @store={"message"=>"Hello", "@version"=>"1", "@timestamp"=>"2015-08-31T12:06:08.273Z", "type"=>"type_1", "host"=>"minix.local"}, @lut={"type"=>[{"message"=>"Hello", "@version"=>"1", "@timestamp"=>"2015-08-31T12:06:08.273Z", "type"=>"type_1", "host"=>"minix.local"}, "type"], "host"=>[{"message"=>"Hello", "@version"=>"1", "@timestamp"=>"2015-08-31T12:06:08.273Z", "type"=>"type_1", "host"=>"minix.local"}, "host"]}>>] {:level=>:warn}
```

I think it is because 'doc_as_upsert' is not sent together with the doc. See part of the log if client logging is enabled:

```
POST http://localhost:9200//_bulk [status:200, request:0.059s, query:0.039s]
> {"update":{"_id":"id_1","_index":"test_index","_type":"type_1","_routing":null}}
> {"doc":{"message":"Hello","@version":"1","@timestamp":"2015-08-31T12:12:40.950Z","type":"type_1","host":"minix.local"}}
```

See https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch.rb#L406
The doc_as_upsert is taken into account only if action is 'update' but my action is set by [@metadata][action]. 

If I remove the 'if' on line 406, I get this request to elastic:

```
POST http://localhost:9200//_bulk [status:200, request:0.083s, query:0.060s]
> {"update":{"_id":"id_1","_index":"test_index","_type":"type_1","_routing":null}}
> {"doc":{"message":"Hello","@version":"1","@timestamp":"2015-08-31T12:13:23.367Z","type":"type_1","host":"minix.local"},"doc_as_upsert":true}
```

It is not proper solution, of course, but at least a hint what's wrong.

Thanks

### PR DESCRIPTION
**Pull request by dchauviere**
This PR fixes a bug (issue #239) in case of interpolated action "update" where specific parameters 'upsert' and 'script' was not set !

### CODE REVIEW

**Comment by suyograo** [comment]
@dchauviere @Da-Wei looks like you used a new GitHub account, so the CLA system did not recognize you. Can you please resign the CLA with this account "@dchauviere"

**Comment by dchauviere** [comment]
@suyograo, I have signed the CLA. Is there a delay before check pass ?

**Comment by suyograo** [comment]
@dchauviere all good now :)

**Comment by dchauviere** [comment]
Thanks @suyograo, its better :)
travis-ci's build seems to randomly fail :
 - previous build failed for ES1.7.5 but succeed for ES2.2.0
 - last build failed for ES2.2.0 but succeed for ES1.7.5

I will try to reproduce ...

**Comment by Sir4ur0n** [comment]
Hi folks,

Is there any progress on this PR please? The issue still happens in Logstash 2.3.2 :'(

Thank you for your contributions!

**Comment by dchauviere** [comment]
@Sir4ur0n @suyograo, with my previous comment, I think we should wait for #437 to be closed before this can be merged, but I have finished to work on it.

**Comment by phoenixgao** [comment]
@dchauviere #437 has been closed for one year, is this going to be merged?

Logshash 5.4.2 still has this issue

**Comment by ItamarBenjamin** [comment]
Hi can someone get this merged please? this bug is really problematic for many developers...

**Comment by dchauviere** [comment]
@suyograo @andrewvc, do you expect anything from me for merging this ?
Should I rebase with master branch ?

**Comment by andrewvc** [comment]
@dchauviere apologies for the delay, I've put it at the top of my todo list tomorrow

**Comment by rjsaran** [comment]
Any plan to merge this in near future.

File: lib/logstash/outputs/elasticsearch/common.rb [hunk 1]
```diff
@@ -10,9 +10,9 @@ module Common
     DOC_CONFLICT_CODE = 409
 
     # When you use external versioning, you are communicating that you want
-    # to ignore conflicts. More obviously, since an external version is a
```
Comment (dchauviere) [inline lib/logstash/outputs/elasticsearch/common.rb]: @guyboertje my editor seems to cleanup unneeded space, if its a problem for merging, I can restore them

**Comment by dchauviere** [comment]
@guyboertje, could you review that your comments in #732 are well integrated here.

File: lib/logstash/outputs/elasticsearch/common.rb (hunk 1 above)
Comment (guyboertje) [reply lib/logstash/outputs/elasticsearch/common.rb]: No need to worry in this case.

**Comment by guyboertje** [comment]
Tests pass. Good to go.

**Comment by ItamarBenjamin** [comment]
Guys thanks for getting this merged! any idea when the next release is due?

**Pull request merged**